crates/alpen-ee/engine/src/sync.rs

```diff
@@ -0,0 +1,1272 @@
+//! Sync missing blocks in execution engine using payloads stored in sequencer database.
+
+use std::future::Future;
+
+use alloy_primitives::B256;
+use alloy_rpc_types_engine::ForkchoiceState;
+use alpen_ee_common::{EnginePayload, ExecBlockStorage, ExecutionEngine, ExecutionEngineError};
+use reth_node_builder::NodeTypesWithDB;
+use reth_provider::{
+    providers::{BlockchainProvider, ProviderNodeTypes},
+    BlockNumReader, ProviderError,
+};
+use strata_acct_types::Hash;
+use thiserror::Error;
+use tracing::{debug, info};
+
+/// Interface for checking if blocks exist in the execution layer.
+///
+/// This abstraction allows for testing without requiring a real Reth provider.
+#[cfg_attr(test, mockall::automock)]
+trait BlockExistenceChecker {
+    /// Check if a block with the given hash exists.
+    fn block_exists(&self, blockhash: Hash) -> Result<bool, SyncError>;
+}
+
+/// Wrapper around [`BlockchainProvider`] that implements [`BlockExistenceChecker`].
+struct RethBlockChecker<'a, N: NodeTypesWithDB + ProviderNodeTypes> {
+    provider: &'a BlockchainProvider<N>,
+}
+
+impl<'a, N: NodeTypesWithDB + ProviderNodeTypes> RethBlockChecker<'a, N> {
+    /// Creates a new Reth block checker.
+    fn new(provider: &'a BlockchainProvider<N>) -> Self {
+        Self { provider }
+    }
+}
+
+impl<'a, N: NodeTypesWithDB + ProviderNodeTypes> BlockExistenceChecker for RethBlockChecker<'a, N> {
+    fn block_exists(&self, blockhash: Hash) -> Result<bool, SyncError> {
+        let b256_hash = B256::from_slice(blockhash.as_ref());
+        Ok(self.provider.block_number(b256_hash)?.is_some())
+    }
+}
+
+/// Errors that can occur during chainstate sync.
+#[derive(Debug, Error)]
+pub enum SyncError {
+    /// Missing exec block at height.
+    #[error("missing exec block at height {0}")]
+    MissingExecBlock(u64),
+
+    /// Missing block payload for specified block hash.
+    #[error("missing block payload for hash {0:?}")]
+    MissingBlockPayload(Hash),
+
+    /// Block was reported as unfinalized but not found in storage.
+    #[error("unfinalized block {0:?} not found in storage")]
+    UnfinalizedBlockNotFound(Hash),
+
+    /// Finalized chain is empty.
+    #[error("finalized chain is empty")]
+    EmptyFinalizedChain,
+
+    /// Storage error.
+    #[error("storage error: {0}")]
+    Storage(#[from] alpen_ee_common::StorageError),
+
+    /// Alpen's execution engine error.
+    #[error("alpen engine error: {0}")]
+    Engine(#[from] ExecutionEngineError),
+
+    /// Reth `Provider` error.
+    #[error("provider error: {0}")]
+    Provider(#[from] ProviderError),
+
+    /// Payload deserialization error.
+    #[error("payload deserialization error: {0}")]
+    PayloadDeserialization(String),
+}
```

Comment:
Can we split this out into a separate module?  Also `error` is redundant in error strings.